### ISSUE TITLE
Hostnames not supported in ring buffer server config line

### ISSUE THREAD

**Issue by jgrammen-agilitypr** [haproxy/haproxy#759]
## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.2.0-1ppa1~focal 2020/07/14 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2025.
Known bugs: http://www.haproxy.org/bugs/bugs-2.2.0.html
Running on: Linux 5.4.0-40-generic #44-Ubuntu SMP Tue Jun 23 00:01:04 UTC 2020 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -O2 -fdebug-prefix-map=/build/haproxy-iKlK9O/haproxy-2.2.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=24).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.3.0
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
              h2 : mode=HTTP       side=FE|BE     mux=H2
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services :
        prometheus-exporter

Available filters :
        [SPOE] spoe
        [COMP] compression
        [TRACE] trace
        [CACHE] cache
        [FCGI] fcgi-app
```
```
Linux [hostname masked] 5.4.0-40-generic #44-Ubuntu SMP Tue Jun 23 00:01:04 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux

```
## What should haproxy do differently? Which functionality do you think we should add?

when adding a server to the server line of a ring buffer hostnames are not supported
```
  ring requests0
      description "haproxy logs"
      format rfc5424
      maxlen 1200
      size 32764
      timeout connect 5s
      timeout server 10s
      server graylog-syslog hostname.internal:4514
```
But the documentation does not indicate this. perhapse either hostnames could be supported just like they are in every other server config line, or the documentation should make it clear that IP is mandatory.

https://cbonte.github.io/haproxy-dconv/2.2/configuration.html#3.9-server
```
Used to configure a syslog tcp server to forward messages from ring buffer.
This supports for all "server" parameters found in 5.2 paragraph.
```
https://cbonte.github.io/haproxy-dconv/2.2/configuration.html#server%20%28Alphabetically%20sorted%20keywords%20reference%29
```<address> is the IPv4 or IPv6 address of the server. Alternatively, a
          resolvable hostname is supported, but this name will be resolved
          during start-up.
```

## What are you trying to do?

A hostname inside of a ring buffer section should work correctly or the documentation should indicate that hostnames are not supported in ring buffers (and peers sections?)

**Comment by wtarreau** [comment]
I think it's a bug, I don't see why they would not work. What error did you get ? I'm suspecting that some late resolution was not involved, but could be wrong. CCing @EmericBr just in case.

**Comment by jgrammen-agilitypr** [comment]
I actually do not receive any error, the logs simply never get sent. and the buffer fills with log messages. upon changing from hostname to IP, the logs immediately start getting sent.

**Comment by wtarreau** [comment]
OK so that sounds pretty much like a resolution that's not done, with logs being sent nowhere. And since for now we don't have any dedicated stats, it's not easy to track indeed!

**Comment by EmericBr** [comment]
Hi, 

"A hostname inside of a ring buffer section should work correctly or the documentation should indicate that hostnames are not supported in ring buffers (and peers sections?)"

Do you notice the same bug on peers sections?

**Comment by jgrammen-agilitypr** [comment]
I do not actually use peers at all, so I cannot comment of whether or not it works. I mentioned peers in the ticket, because when I asked for help in the haproxy slack the person who replied indicated that it didn't work in either peers or ring buffers

```
jethro (IRC)APP  5:52 PM
@Jason Grammenos haproxy will only resolve hostnames on server lines in backend or listen blocks. for peers or ring sections you have to use IP addresses.
```

**Comment by EmericBr** [comment]
[0001-BUG-MEDIUM-resolve-fix-init-resolving-for-ring-and-p.txt](https://github.com/haproxy/haproxy/files/4954351/0001-BUG-MEDIUM-resolve-fix-init-resolving-for-ring-and-p.txt)
patch proposal to test

**Comment by jgrammen-agilitypr** [comment]
if the patch proposal above was intended for me to use to test, I apologize but I do not have the ability to compile haproxy, I am running from ppa based builds.

**Comment by wtarreau** [comment]
You don't have to apologize, don't worry, we'll take care of it then.

**Comment by EmericBr** [comment]
Don't worry, and thanks for the report !

**Comment by jmagnin** [comment]
The patch fixes resolution when names are used on server lines for peers or ring sections. 
it's worth noting that if the name is not resolvable haproxy will refuse to start.

**Comment by wtarreau** [comment]
That's indeed the intent as it's already what is done in other cases ("peer" instead of "server" or "log" instead of "server"). Thanks for the tests Jérôme!

**Comment by wtarreau** [comment]
I've now merged Emeric's patch and backported the peers-related part as far as 2.0, so this will be fixed in 2.2.1 once it's out. I'm requalifying it as a bug (as we really broke peers in 2.0) and closing it as resolved.

Thanks everyone!
